File: Lab5/pzpi-22-2-dobrovolska-mariia-lab5/models/Alert.js
const mongoose = require('mongoose');

const alertSchema = new mongoose.Schema({
  deviceId: {
    type: String,
    required: true,
    index: true
  },
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },
  type: {
    type: String,
    enum: ['TEMPERATURE', 'HUMIDITY', 'DEVICE_OFFLINE', 'SENSOR_ERROR'],
    required: true
  },
  severity: {
    type: String,
    enum: ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL'],
    default: 'MEDIUM'
  },
  message: {
    type: String,
    required: true
  },
  currentValue: {
    type: Number
  },
  targetValue: {
    type: Number
  },
  isResolved: {
    type: Boolean,
    default: false
  },
  resolvedAt: {
    type: Date
  },
  emailSent: {
    type: Boolean,
    default: false
  },
  emailSentAt: {
    type: Date
  },
  metadata: {
    temperature: Number,
    humidity: Number,
    lightLevel: Number
  }
}, {
  timestamps: true
});

// Index for efficient querying
alertSchema.index({ deviceId: 1, createdAt: -1 });
alertSchema.index({ userId: 1, isResolved: 1, createdAt: -1 });

// TTL index to automatically delete resolved alerts after 90 days
alertSchema.index({ resolvedAt: 1 }, { expireAfterSeconds: 7776000, partialFilterExpression: { isResolved: true } });

module.exports = mongoose.model('Alert', alertSchema);

File: Lab4/pzpi-22-2-dobrovolska-mariia-lab4/app/src/main/java/com/example/apz_lab4/models/Alert.kt
// models/Alert.kt
package com.example.apz_lab4.models

data class Alert(
    val _id: String,
    val deviceId: String,
    val userId: String,
    val type: String,
    val severity: String,
    val message: String,
    val currentValue: Any,
    val isResolved: Boolean,
    val emailSent: Boolean,
    val createdAt: String,
    val updatedAt: String,
    val deviceName: String?, // nullable — бо не всі мають
    val metadata: AlertMetadata
)

data class AlertMetadata(
    val temperature: Double,
    val humidity: Double,
    val lightLevel: Int
)


